src/dataflow/src/sink/kafka.rs

```diff
@@ -187,11 +187,21 @@ impl SinkMetrics {
 
 pub struct SinkProducerContext {
     metrics: Arc<SinkMetrics>,
+    retry_queue: Arc<Mutex<VecDeque<OwnedMessage>>>,
```

Comment:
Happy to pull these out to a separate `KafkaSinkProducerRetryQueue` that encapsulates some of the semantics more -- to do things like ensure things are read in the right order.

I'd have to think about it more to be sure but it might be that correctness in this PR relies on 1) on error: adding a message to the retry queue and then decrementing the outstanding send out; and 2) when flushing: checking the send count and then the retry queue side

Reply:
I think it might be cleaner and more performant to just use a (unbounded, as we like unbounded queues in materialized right now) crossbeam channel!

Reply:
I thought about doing that -- and it seemed better to have this `VecDeque` than to have to write extra code to handle the hypothetical [Disconnected](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/enum.TryRecvError.html) error that we want to be impossible

what do you thinK?